src/VisualStudio/Core/Def/Implementation/LanguageClient/LanguageServerClient.cs

```diff
@@ -103,38 +97,31 @@ public async Task<Connection> ActivateAsync(CancellationToken cancellationToken)
-        /// To start the server, invoke the <see cref="StartAsync"/> event;
+        /// Signals that the extension has been loaded.
+        /// The caller expects that <see cref="ActivateAsync(CancellationToken)"/> can be called
+        /// immediately following the completion of this method.
         /// </summary>
-        public Task OnLoadedAsync()
+        public async Task OnLoadedAsync()
         {
-            var token = _asyncListener.BeginAsyncOperation("OnLoadedAsync");
+            // initialize things on UI thread
+            await InitializeOnUIAsync().ConfigureAwait(false);
 
-            // set up event stream so that we start LSP server once Roslyn is loaded
-            _eventListener.WorkspaceStarted.ContinueWith(async _ =>
+            // this might get called before solution is fully loaded and before file is opened. 
+            // we delay our OOP start until then, but user might do vsstart before that. so we make sure we start OOP if 
+            // it is not running yet. multiple start is no-op
+            ((RemoteHostClientServiceFactory.RemoteHostClientService)_workspace.Services.GetService<IRemoteHostClientService>()).Enable();
+
+            // wait until remote host is available before let platform know that they can activate our LSP
+            var client = await _workspace.TryGetRemoteHostClientAsync(CancellationToken.None).ConfigureAwait(false);
+            if (client == null)
             {
-                // initialize things on UI thread
-                await InitializeOnUIAsync().ConfigureAwait(false);
-
-                // this might get called before solution is fully loaded and before file is opened. 
-                // we delay our OOP start until then, but user might do vsstart before that. so we make sure we start OOP if 
-                // it is not running yet. multiple start is no-op
-                ((RemoteHostClientServiceFactory.RemoteHostClientService)_workspace.Services.GetService<IRemoteHostClientService>()).Enable();
-
-                // wait until remote host is available before let platform know that they can activate our LSP
-                var client = await _workspace.TryGetRemoteHostClientAsync(CancellationToken.None).ConfigureAwait(false);
-                if (client == null)
-                {
-                    // there is no OOP. either user turned it off, or process got killed.
-                    // don't ask platform to start LSP
-                    return;
-                }
-
-                // let platform know that they can start us
-                await StartAsync.InvokeAsync(this, EventArgs.Empty).ConfigureAwait(false);
-            }, TaskScheduler.Default).CompletesAsyncOperation(token);
+                // there is no OOP. either user turned it off, or process got killed.
+                // don't ask platform to start LSP
+                return;
```

Comment:
Would this be worth throwing up a gold bar here?  Because clearly we're in a bad state here.  Can definitely also add a watson for that.
@heejaechang will we already throw up a gold bar if this happens?

Reply:
IMO, this should be up to the LSP client and/or API to handle.

That said, I'd hope we wouldn't find ourselves in this scenario anymore.

Reply:
We can definitely still end up here if the user kills the OOP / OOP crashes as the original comment.

also adding @tinaschrepfer also to comment - is throwing here the appropriate solution, or do we need to take care of displaying the exception properly on our end

Reply:
it depends on what we want. if client is null, it means something is broken. and we should have reacted to it when it happened (info bar/NFW and etc) but whether we can connect that to this particular code is different story since the issue could happen 10 mins before and user kept using VS even if info bar told users to close and re-open VS.

so, one could use session explorer to see whether we have got into broken state before this code took place, or one could get dump to see whether OOP of roslyn is turned off, or we lost connection to OOP, or add telemetry to this code path to indicate that we took this code path.

all depends on what you are trying to do.

Reply:
I see, thanks!  I think the only solution is to restart VS, so since we already display a gold bar indicating that they should restart VS there's no need to show another one here specifically.  Probably best to just log telemetry here that we hit this then.

Reply:
Throwing exceptions here can have ramifications further down the line.  We don't expect this method to throw exceptions, so it cause a cascading effect when you do throw and potentially deliver a poor user experience.  Our assumption is that in this method, you control if you actually want to activate the language server.

Reply:
>We don't expect this method to throw exceptions, so it cause a cascading effect when you do throw and potentially deliver a poor user experience.

The problem I can see with requiring the language service to do their own alerting the user of the failure condition here is the behavior is going to be inconsistent and fragmented across languages. Some will give dialogs, some gold bars, some will do nothing.

Editor has guarded operations as a generic catch-all and reporting mechanism for exceptions. I was advocating for such a model because it keeps things consistent. If the LS is dying anyways, at least the experience is consistent.